bin/parse_kraken2report.py

```diff
@@ -0,0 +1,168 @@
+            keys.extend(getAllKeys(v, newKey, sep=sep))
+
+    return keys
+
+def removeIncompleteTaxa(listKrakenTaxonomicKeys, sep = '\t'):
+    # remove the incomplete paths to the leafs of the hierarchical tree, keep all paths that are ending in a leaf
+    listKrakenTaxonomicKeys.sort()
+    to_remove = set()
+
+    for i in range(len(listKrakenTaxonomicKeys) - 1):
+        if listKrakenTaxonomicKeys[i + 1].startswith(listKrakenTaxonomicKeys[i] + sep):
+            to_remove.add(i)
+
+    return [item for idx, item in enumerate(listKrakenTaxonomicKeys) if idx not in to_remove]
+
+def generateDictForLookupOfTaxonomicSubentries(removedIncompleteTaxaList):
+    # This function generates a lookup dictionary that allows to get all subentries
+    # of a certain taxonomic entry (e.g. for 'Homo' it looks as follows:
+    # {'Homo':'Homo sapiens'})
+    lookupDict = {}
+
+    for entry in removedIncompleteTaxaList:
+        listOfTaxonomicNames = entry.split('\t')
+
+        while listOfTaxonomicNames:
+            newTaxomomicName = listOfTaxonomicNames.pop(0)
+
+            if newTaxomomicName in lookupDict:
+                lookupDict[newTaxomomicName].update(listOfTaxonomicNames)
+            else:
+                lookupDict[newTaxomomicName] = set(listOfTaxonomicNames)
+
+    for key in lookupDict:
+        lookupDict[key] = sorted(list(lookupDict[key]))
+
+    sortedKeys = sorted(lookupDict)
+    return {k: lookupDict[k] for k in sortedKeys}
+
+def getKeysWithIDs(krakenHierarchicalTaxonomyDict, results=None):
+    # itterate over the hierarchical taxonomy dictionary and create
+    # a mapping from taxonomic name to taxonomic id.
+    # This is done recursively and results=None is the inital state.
+    if results is None:
+        results = {}
+
+    for k, v in krakenHierarchicalTaxonomyDict.items():
+        if isinstance(v, dict):
+            if 'id' in v:
+                results[k] = v['id']
+            results.update(getKeysWithIDs(v, results))
+    return results
+
+def main():
+
+    args = parse_args()
+
+    krakenList = read_in_kraken2report(args.input)
+    krakenHierarchicalTaxonomyDict = krakenTaxonomy2hierarchy(krakenList)
+    listKrakenTaxonomicKeys = getAllKeys(krakenHierarchicalTaxonomyDict)
+    resultRemovedIncompleteTaxa = removeIncompleteTaxa(listKrakenTaxonomicKeys)
+    resultLookUpDict = generateDictForLookupOfTaxonomicSubentries(resultRemovedIncompleteTaxa)
+    try:
+        # Try if the taxon to assess for/to filter is in the database (represented by the lookup dictionary)
+        # and add all subentries of the 'tax2filter' plus the 'tax2filter' to a single list
+        resultToFilter = resultLookUpDict[args.tax2filter] + [args.tax2filter]
+    except KeyError:
+        raise KeyError("The taxaomic group/taxon you want to check for/filter out is not in the kraken database. Use a database that includes the taxonomic group or taxon or change the tax2filter parameter to something that is in the database.")
```

Comment:
Will raising `KeyError` cause the pipeline to fail loudly? (I guess it should.)

Reply:
Yes, it will!